server/channels/app/plugin_hooks_test.go

```diff
@@ -1632,3 +1636,74 @@ func TestHookMessagesWillBeConsumed(t *testing.T) {
 		assert.Equal(t, "mwbc_plugin:message", post.Message)
 	})
 }
+
+func TestHookPreferencesHaveChanged(t *testing.T) {
+	t.Run("should be called when preferences are changed by non-plugin code", func(t *testing.T) {
+		th := Setup(t).InitBasic()
+		defer th.TearDown()
+
+		// Setup plugin
+		var mockAPI plugintest.API
+
+		tearDown, pluginIDs, _ := SetAppEnvironmentWithPlugins(t, []string{`
+			package main
+
+			import (
+				"github.com/mattermost/mattermost/server/public/plugin"
+				"github.com/mattermost/mattermost/server/public/model"
+			)
+
+			type MyPlugin struct {
+				plugin.MattermostPlugin
+			}
+
+			func (p *MyPlugin) PreferencesHaveChanged(c *plugin.Context, preferences []*model.Preference) {
+				for _, preference := range preferences {
+					p.API.LogDebug(fmt.Sprintf("category=%s name=%s value=%s", preference.Category, preference.Name, preference.Value))
+				}
+			}
+
+			func main() {
+				plugin.ClientMain(&MyPlugin{})
+			}
+		`}, th.App, func(*model.Manifest) plugin.API { return &mockAPI })
+		defer tearDown()
+
+		// Confirm plugin is actually running
+		require.Len(t, pluginIDs, 1)
+		pluginID := pluginIDs[0]
+
+		require.True(t, th.App.GetPluginsEnvironment().IsActive(pluginID))
+
+		// Setup test
+		preferences := model.Preferences{
+			{
+				UserId:   th.BasicUser.Id,
+				Category: "test_category",
+				Name:     "test_name_1",
+				Value:    "test_value_1",
+			},
+			{
+				UserId:   th.BasicUser.Id,
+				Category: "test_category",
+				Name:     "test_name_2",
+				Value:    "test_value_2",
+			},
+		}
+
+		mockAPI.On("LogDebug", "category=test_category name=test_name_1 value=test_value_1")
+		mockAPI.On("LogDebug", "category=test_category name=test_name_2 value=test_value_2")
+		defer mockAPI.AssertExpectations(t)
+
+		// Run test
+		err := th.App.UpdatePreferences(th.Context, th.BasicUser.Id, preferences)
+
+		require.Nil(t, err)
+
+		// Hooks are run in a goroutine, so wait for those to complete
+		time.Sleep(1 * time.Second)
```

Comment:
I mentioned this in the server channel, but it feels like it would be nice if we could wait for goroutines to finish instead of adding an arbitrary timeout to these tests

Reply:
Does the hook need to run in a goroutine in the first place?

Reply:
It looks like we always run hooks that happen after an event occurs in a goroutine since we don't need to block on their results

Reply:
Sure, still I wonder how much we are really saving by running it asynchronously. I guess semantically it makes sense if we don't want plugins to potentially alter the response. But I think there could be valid use cases for that such as preventing or overriding some preferences from the plugin side but I assume that's not what we require today.

Reply:
If we wanted that, I think we'd want a separate "will" hook for that since that'd need to happen before the preference is saved to the DB